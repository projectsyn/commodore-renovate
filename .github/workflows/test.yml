name: Test
on:
  pull_request:
    branches:
      - master

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '18'
          cache: yarn
      - name: Installing dependencies
        run: yarn install --frozen-lockfile
      - name: Lint
        run: |
          yarn lint
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '18'
          cache: yarn
      - name: Installing dependencies
        run: yarn install --frozen-lockfile
      - name: Setup Python
        uses: actions/setup-python@v4
        with:
          python-version: '3.9'
      - name: Cache pip
        uses: actions/cache@v3
        with:
          # This path is specific to Ubuntu
          path: ~/.cache/pip
          # Look to see if there is a cache hit for the corresponding requirements file
          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
          restore-keys: |
            ${{ runner.os }}-pip-
            ${{ runner.os }}-
      - name: Install dependencies
        run: pip install -r requirements.txt
      - name: Test
        run: |
          yarn test

  docker-build:
    runs-on: ubuntu-latest
    env:
      IMAGE: ${{ github.repository }}
    steps:
      - uses: actions/checkout@v4
      - name: Get changed files
        id: changed-files
        uses: tj-actions/changed-files@v39
      - name: Build
        if: contains(steps.changed-files.outputs.modified_files, 'Dockerfile')
        uses: docker/build-push-action@v4
        with:
          push: false
          tags: 'ghcr.io/${{ env.IMAGE }}:test'
