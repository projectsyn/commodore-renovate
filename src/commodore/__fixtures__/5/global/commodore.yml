classes:
  - global.params
  - global.${facts:distribution}
  - ${cluster:tenant}.${cluster:name}
