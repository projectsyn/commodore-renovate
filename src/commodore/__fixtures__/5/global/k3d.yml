parameters:
  components:
    argocd:
      url: https://github.com/projectsyn/component-argocd-k3d.git
