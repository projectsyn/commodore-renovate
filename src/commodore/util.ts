import clone from 'just-clone';
import yaml from 'js-yaml';
import Git from 'simple-git';

import { parse } from 'path';
import { mkdir, readFile, writeFile } from 'fs/promises';
import { URL } from 'url';

import { logger } from 'renovate/dist/logger';
import { getGlobalConfig } from 'renovate/dist/config/global';
import { simpleGitConfig } from 'renovate/dist/util/git/config';

import type { Facts, RepoConfig } from './types';

export function cacheDir(): string {
  let cacheDir: string | undefined = getGlobalConfig().cacheDir;
  if (cacheDir === undefined) {
    cacheDir = '/tmp/renovate';
  }
  return cacheDir;
}

/* Prune object, removing any null or undefined keys */
export function pruneObject(obj: Facts) {
  return Object.entries(obj)
    .filter(([_, v]) => v !== undefined && v !== null)
    .reduce((main, [key, value]) => ({ ...main, [key]: value }), {});
}

/* Render object as YAML and write to file */
export async function writeYamlFile(
  filePath: string,
  obj: object
): Promise<void> {
  const objYaml: string = yaml.dump(obj);
  return writeFile(filePath, objYaml);
}

/* Extract distribution/cloud/region from filename */
export function parseFileName(
  fileName: string,
  distributionRegex: RegExp,
  cloudRegionRegex: RegExp,
  ignoreValues: Array<string>
): Facts {
  const distA = fileName.match(distributionRegex);
  const cloudA = fileName.match(cloudRegionRegex);

  const dist = distA ? distA.groups : null;
  const cloud = cloudA ? cloudA.groups : null;

  let facts: Facts = {
    distribution: null,
    cloud: null,
    region: null,
  };
  if (dist !== undefined && dist !== null) {
    facts.distribution = dist.distribution;
    facts.cloud = dist.cloud ? dist.cloud : null;
  }
  if (cloud !== undefined && cloud !== null) {
    facts.cloud = cloud.cloud;
    facts.region = cloud.region ? cloud.region : null;
  }

  if (
    facts.distribution !== null &&
    ignoreValues.includes(facts.distribution)
  ) {
    facts.distribution = null;
  }
  if (facts.cloud !== null && ignoreValues.includes(facts.cloud)) {
    facts.cloud = null;
  }
  if (facts.region !== null && ignoreValues.includes(facts.region)) {
    facts.region = null;
  }

  return facts;
}

function isObject(item: any) {
  return item && typeof item === 'object' && !Array.isArray(item);
}

// Adapted from https://stackoverflow.com/a/37164538
export function mergeConfig(base: any, config: any): any {
  let output = clone(base);
  if (isObject(base) && isObject(config)) {
    Object.keys(config).forEach((key) => {
      if (isObject(config[key])) {
        if (!(key in base)) Object.assign(output, { [key]: config[key] });
        else output[key] = mergeConfig(base[key], config[key]);
      } else if (Array.isArray(config[key])) {
        if (!(key in base)) {
          Object.assign(output, { [key]: Array.from(config[key]) });
        } else if (Array.isArray(output[key])) {
          output[key].push(...config[key]);
        } else {
        }
      } else {
        Object.assign(output, { [key]: config[key] });
      }
    });
  }
  return output;
}

export var globalRepos: Map<string, RepoConfig> = new Map();

export function globalRepoDir(globalRepoURL: string): string {
  let repoUrl = new URL(globalRepoURL);
  let repoPathInfo = parse(repoUrl.pathname);
  let repoPath = `${repoUrl.host}${repoPathInfo.dir}/${repoPathInfo.name}`;
  return cacheDir() + `/global-repos/${repoPath}`;
}

export async function cloneGlobalRepo(config: any): Promise<RepoConfig> {
  if (globalRepos.has(config.globalRepoURL)) {
    return globalRepos.get(config.globalRepoURL) as RepoConfig;
  }

  logger.info(
    {
      globalRepoURL: config.globalRepoURL,
      tenantId: config.tenantId,
    },
    'Global repo for this tenant is not initialized yet, cloning it'
  );
  const dir: string = globalRepoDir(config.globalRepoURL);
  await mkdir(dir, { recursive: true });
  // TODO(sg): support Git-https?
  const git: any = Git(dir, simpleGitConfig());
  await git.clone(config.globalRepoURL, '.');
  const globalExtraConfig = await parseGlobalRepoConfig(dir);
  const rc: RepoConfig = {
    dir: dir,
    extraConfig: globalExtraConfig,
  } as RepoConfig;
  globalRepos.set(config.globalRepoURL, rc);
  return rc;
}

export async function parseGlobalRepoConfig(globalDir: string): Promise<any> {
  const globalDirConfigFile: string = globalDir + '/renovate.json';
  let globalDirConfig: any = {};
  try {
    const globalConfigStr: string = (
      await readFile(globalDirConfigFile)
    ).toString();
    globalDirConfig = JSON.parse(globalConfigStr) as any;
  } catch (e: any) {
    logger.info(
      {
        globalDir: globalDir,
        exception: e,
      },
      'Exception while reading global repo config'
    );
  }
  if (
    globalDirConfig.commodore != null &&
    globalDirConfig.commodore.extraConfig != null
  ) {
    try {
      const globalExtraConfigStr = (
        await readFile(globalDir + '/' + globalDirConfig.commodore.extraConfig)
      ).toString();
      return JSON.parse(globalExtraConfigStr) as any;
    } catch (e: any) {
      logger.info(
        {
          globalDir: globalDir,
          extraConfig: globalDirConfig.commodore.extraConfig,
          exception: e,
        },
        'Exception while reading global repo extraConfig'
      );
    }
  }
  return {};
}
