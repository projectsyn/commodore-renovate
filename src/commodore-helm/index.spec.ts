import { getFixturePath } from '../test/util';
import { extractPackageFile, extractAllPackageFiles, extractHelmChartDependencies } from './index';
import { beforeEach, expect, describe, it } from '@jest/globals';

import { getGlobalConfig } from 'renovate/dist/config/global';

import type { BunyanRecord } from 'renovate/dist/logger/types';
import { ERROR } from 'bunyan';

import { clearProblems, getProblems } from 'renovate/dist/logger';

jest.mock('renovate/dist/config/global');
function mockGetGlobalConfig(
  fixtureId: string,
): void {
  const mockGetGlobalConfigFn = getGlobalConfig as jest.MockedFunction<
    typeof getGlobalConfig
  >;
  mockGetGlobalConfigFn.mockImplementation(() => {
    return {
      localDir: getFixturePath(fixtureId),
      cacheDir: '/tmp/renovate',
    };
  });
}

function getLoggerErrors(): BunyanRecord[] {
  const loggerErrors = getProblems().filter((p) => p.level >= ERROR);
  return loggerErrors;
}

beforeEach(() => {
  // clear logger
  clearProblems();
});


describe('manager/commodore-helm/index', () => {
  describe('extractPackageFile()', () => {
  });
  describe('extractAllPackageFiles()', () => {
    it('extracts standard Helm dependencies', async () => {
      mockGetGlobalConfig('1');
      const res = await extractAllPackageFiles({}, ['class/defaults.yml', 'class/component-name.yml']);
      const errors = getLoggerErrors();
      if (errors.length > 0) {
        console.log(errors);
      }
      expect(errors.length).toBe(0);
      expect(res).not.toBeNull();
      if (res) {
        expect(res.length).toBe(1);
        if (res.length == 1) {
          const res0 = res[0];
          expect(res0.packageFile).toBe('class/defaults.yml');
          const deps = res0.deps;
          expect(deps).toMatchSnapshot();
          expect(deps.length).toBe(2);
        }
      }
    });
    it('extracts standard Helm dependencies for components with long names', async () => {
      mockGetGlobalConfig('2');
      const res = await extractAllPackageFiles({}, ['class/defaults.yml', 'class/long-component-name.yml']);
      const errors = getLoggerErrors();
      if (errors.length > 0) {
        console.log(errors);
      }
      expect(errors.length).toBe(0);
      expect(res).not.toBeNull();
      if (res) {
        expect(res.length).toBe(1);
        if (res.length == 1) {
          const res0 = res[0];
          expect(res0.packageFile).toBe('class/defaults.yml');
          const deps = res0.deps;
          expect(deps).toMatchSnapshot();
          expect(deps.length).toBe(2);
        }
      }
    });
    it('extracts Helm dependencies with mismatched names', async () => {
      mockGetGlobalConfig('3');
      const res = await extractAllPackageFiles({}, ['class/defaults.yml', 'class/component-name.yml']);
      const errors = getLoggerErrors();
      if (errors.length > 0) {
        console.log(errors);
      }
      expect(errors.length).toBe(0);
      expect(res).not.toBeNull();
      if (res) {
        expect(res.length).toBe(1);
        if (res.length == 1) {
          const res0 = res[0];
          expect(res0.packageFile).toBe('class/defaults.yml');
          const deps = res0.deps;
          expect(deps).toMatchSnapshot();
          expect(deps.length).toBe(2);
        }
      }
    });
  });
  describe('extractHelmChartDependencies()', () => {
  });
});
