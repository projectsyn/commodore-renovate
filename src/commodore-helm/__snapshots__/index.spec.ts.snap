// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager/commodore-helm/index extractAllPackageFiles() extracts new and old standard Helm dependencies in the same file 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
    "groupName": "component-name",
    "registryUrls": [
      "https://charts.example.com/",
    ],
  },
  {
    "currentValue": "4.5.6",
    "depName": "chart-2",
    "groupName": "component-name",
    "registryUrls": [
      "https://charts2.example.com/",
    ],
  },
]
`;

exports[`manager/commodore-helm/index extractAllPackageFiles() extracts old standard Helm dependencies 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
    "groupName": "component-name",
    "registryUrls": [
      "https://charts.example.com/",
    ],
  },
  {
    "currentValue": "4.5.6",
    "depName": "chart-2",
    "groupName": "component-name",
    "registryUrls": [
      "https://charts2.example.com/",
    ],
  },
]
`;

exports[`manager/commodore-helm/index extractAllPackageFiles() extracts old standard Helm dependencies for components with long names 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
    "groupName": "long-component-name",
    "registryUrls": [
      "https://charts.example.com/",
    ],
  },
  {
    "currentValue": "4.5.6",
    "depName": "chart-2",
    "groupName": "long-component-name",
    "registryUrls": [
      "https://charts2.example.com/",
    ],
  },
]
`;

exports[`manager/commodore-helm/index extractAllPackageFiles() extracts old-style Helm dependencies with mismatched names 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
    "groupName": "component-name",
    "propSource": "chart_1",
    "registryUrls": [
      "https://charts.example.com/",
    ],
  },
  {
    "currentValue": "4.5.6",
    "depName": "chart-2",
    "groupName": "component-name",
    "propSource": "chart2",
    "registryUrls": [
      "https://charts2.example.com/",
    ],
  },
]
`;

exports[`manager/commodore-helm/index extractAllPackageFiles() gracefully ignores components with old standard \`charts\` parameter but no Kapitan config 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
    "groupName": "component-name",
    "skipReason": "invalid-dependency-specification",
  },
]
`;

exports[`manager/commodore-helm/index extractAllPackageFiles() gracefully ignores components with old standard \`charts\` parameter but no Kapitan helm dependencies 1`] = `
[
  {
    "currentValue": "4.5.6",
    "depName": "chart-1",
    "groupName": "component-name-2",
    "skipReason": "invalid-dependency-specification",
  },
]
`;

exports[`manager/commodore-helm/index extractPackageFile() extracts Helm chart versions for mismatched keys when called with sufficient config 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
  },
  {
    "currentValue": "4.5.6",
    "depName": "chart-2",
  },
]
`;

exports[`manager/commodore-helm/index extractPackageFile() extracts Helm chart versions from new and old standard 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
    "registryUrls": [
      "https://charts.example.com/",
    ],
  },
  {
    "currentValue": "4.5.6",
    "depName": "chart-2",
  },
]
`;

exports[`manager/commodore-helm/index extractPackageFile() extracts Helm chart versions when called with sufficient config 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
  },
  {
    "currentValue": "4.5.6",
    "depName": "chart-2",
  },
]
`;

exports[`manager/commodore-helm/index extractPackageFile() handles wrong dependency info gracefully 1`] = `
[
  {
    "currentValue": "1.2.3",
    "depName": "chart-1",
  },
  {
    "currentValue": "4.5.6",
    "skipReason": "invalid-dependency-specification",
  },
]
`;
